File: android/LocalBridge.Android/app/src/main/java/com/localbridge/android/core/protocol/ProtocolEnvelopeValidator.kt
package com.localbridge.android.core.protocol

data class EnvelopeValidationResult(
    val isValid: Boolean,
    val errorCode: String? = null,
    val errorMessage: String? = null
) {
    companion object {
        val Valid = EnvelopeValidationResult(isValid = true)
    }
}

object ProtocolEnvelopeValidator {
    fun <T> validate(
        envelope: ProtocolEnvelope<T>?,
        requirePayload: Boolean = true,
        vararg expectedPacketTypes: String
    ): EnvelopeValidationResult {
        if (envelope == null) {
            return EnvelopeValidationResult(
                isValid = false,
                errorCode = ProtocolErrorCodes.invalidRequest,
                errorMessage = "Protocol envelope is missing."
            )
        }

        if (envelope.meta.version != ProtocolConstants.version) {
            return EnvelopeValidationResult(
                isValid = false,
                errorCode = ProtocolErrorCodes.protocolMismatch,
                errorMessage = "Protocol version mismatch."
            )
        }

        if (expectedPacketTypes.isNotEmpty() && expectedPacketTypes.none { it == envelope.meta.packetType }) {
            return EnvelopeValidationResult(
                isValid = false,
                errorCode = ProtocolErrorCodes.invalidRequest,
                errorMessage = "Unexpected packet type ${envelope.meta.packetType}."
            )
        }

        if (requirePayload && envelope.payload == null) {
            return EnvelopeValidationResult(
                isValid = false,
                errorCode = ProtocolErrorCodes.invalidRequest,
                errorMessage = "Protocol payload is missing."
            )
        }

        return EnvelopeValidationResult.Valid
    }
}

File: src/LocalBridge.Core/Protocol/ProtocolEnvelopeValidator.cs
namespace LocalBridge.Core.Protocol;

public sealed record ProtocolEnvelopeValidationResult(
    bool IsValid,
    string? ErrorCode = null,
    string? ErrorMessage = null)
{
    public static ProtocolEnvelopeValidationResult Valid { get; } = new(true);
}

public static class ProtocolEnvelopeValidator
{
    public static ProtocolEnvelopeValidationResult Validate<TPayload>(
        ProtocolEnvelope<TPayload>? envelope,
        bool requirePayload = true,
        params string[] expectedPacketTypes)
    {
        if (envelope is null || envelope.Meta is null)
        {
            return Invalid(
                ProtocolErrorCodes.InvalidRequest,
                "Protocol envelope is missing or malformed.");
        }

        if (!string.Equals(envelope.Meta.Version, ProtocolConstants.Version, StringComparison.OrdinalIgnoreCase))
        {
            return Invalid(
                ProtocolErrorCodes.ProtocolMismatch,
                "Protocol version mismatch.");
        }

        if (expectedPacketTypes.Length > 0 &&
            !Array.Exists(
                expectedPacketTypes,
                packetType => string.Equals(packetType, envelope.Meta.PacketType, StringComparison.OrdinalIgnoreCase)))
        {
            return Invalid(
                ProtocolErrorCodes.InvalidRequest,
                $"Unexpected packet type '{envelope.Meta.PacketType}'.");
        }

        if (requirePayload && envelope.Payload is null)
        {
            return Invalid(
                ProtocolErrorCodes.InvalidRequest,
                "Protocol payload is missing.");
        }

        return ProtocolEnvelopeValidationResult.Valid;
    }

    private static ProtocolEnvelopeValidationResult Invalid(string errorCode, string errorMessage)
    {
        return new ProtocolEnvelopeValidationResult(
            IsValid: false,
            ErrorCode: errorCode,
            ErrorMessage: errorMessage);
    }
}
